components/script/dom/bindings/typedarrays.rs

```diff
@@ -0,0 +1,112 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::ptr;
+
+use js::jsapi::{Heap, JSObject, JS_GetArrayBufferViewBuffer};
+use js::rust::wrappers::DetachArrayBuffer;
+use js::rust::CustomAutoRooterGuard;
+use js::typedarray::{CreateWith, Float32Array};
+
+use crate::script_runtime::JSContext;
+
+#[derive(JSTraceable)]
+pub struct Float32ArrayOnHeap {
+    internal: Heap<*mut JSObject>,
+}
+
+impl Float32ArrayOnHeap {
+    pub fn default() -> Self {
+        Float32ArrayOnHeap {
+            internal: Heap::default(),
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn set_data(&self, cx: JSContext, data: &[f32]) -> Result<(), ()> {
+        rooted!(in (*cx) let mut array = ptr::null_mut::<JSObject>());
+
+        let res =
+            unsafe { Float32Array::create(*cx, CreateWith::Slice(&data), array.handle_mut()) };
+
+        if res.is_err() {
+            return Err(());
+        }
+
+        self.internal.set(*array);
+        Ok(())
+    }
+
+    #[allow(unsafe_code)]
+    pub fn acquire_data(&self, cx: JSContext) -> Result<Vec<f32>, ()> {
+        typedarray!(in(*cx) let array: Float32Array = self.internal.get());
+        let data = if let Ok(array) = array {
+            let data = array.to_vec();
+            let mut is_shared = false;
+            unsafe {
+                rooted!(in (*cx) let view_buffer =
+                    JS_GetArrayBufferViewBuffer(*cx, self.internal.handle(), &mut is_shared));
+                // This buffer is always created unshared
+                debug_assert!(!is_shared);
+                let _ = DetachArrayBuffer(*cx, view_buffer.handle());
+            }
+            Ok(data)
+        } else {
+            Err(())
+        };
+        self.internal.set(ptr::null_mut());
+        data
+    }
+
+    #[allow(unsafe_code)]
+    pub fn copy_data_to(
+        &self,
+        cx: JSContext,
+        dest: &mut [f32],
+        start: usize,
+        end: usize,
+    ) -> Result<(), ()> {
+        assert!(self.is_set());
+        typedarray!(in(*cx) let array: Float32Array = self.internal.get());
+        if array.is_err() {
+            return Err(());
+        }
+        let array = array.unwrap();
+        unsafe {
+            let slice = (*array).as_slice();
+            dest.clone_from_slice(&slice[start..end]);
+        }
+        Ok(())
+    }
+
+    #[allow(unsafe_code)]
+    pub fn copy_data_from(
+        &self,
+        cx: JSContext,
+        source: CustomAutoRooterGuard<Float32Array>,
+        start: usize,
+        end: usize,
```

Comment:
These arguments surprised me, because start is only used on the destination array, not on the source array.

Reply:
I think `end` could be removed if `AudioBuffer::length` was passed to the typed array on initialization instead of just calling `default`, but since the length is a constraint that is part of the web API, perhaps it should be left as is? Is that what you mean?

Reply:
My confusion is that as an API for interacting with two typed arrays, there's no indication that start is only used for the destination. If there were dest_start and length arguments, that would be more clear.

Reply:
Ok I see, I'll rename those as such then.